pkg/ipam/poolallocator/allocator.go

```diff
@@ -426,6 +520,51 @@ func (a *IPPoolAllocator) ReleasePod(namespace, podName string) error {
 	return err
 }
 
+// ReleaseResource releases all IPs associated with specified StatefulSet. It returns error
+// in case CRD failed to update its state or if pool or StatefulSet are not present.
+// In case of success, IP pool CRD status is updated with released entries.
+func (a *IPPoolAllocator) ReleaseStatefulSet(namespace, name string) error {
+
+	// Retry on CRD update conflict which is caused by multiple agents updating a pool at same time.
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		ipPool, err := a.getPool()
+
+		if err != nil {
+			return err
+		}
+
+		var updatedAdresses []v1alpha2.IPAddressState
+		for _, ip := range ipPool.Status.IPAddresses {
+			if ip.Owner.StatefulSet == nil || ip.Owner.StatefulSet.Namespace != namespace || ip.Owner.StatefulSet.Name != name {
+				updatedAdresses = append(updatedAdresses, ip)
+			}
+		}
+
+		if len(ipPool.Status.IPAddresses) == len(updatedAdresses) {
+			// no change
+			klog.InfoS("No reserved IPs found", "pool", ipPool.Name, "Namespace", namespace, "StatefulSet", name)
```

Comment:
too verbose to log such information in default level. Please check other logs introduced by this PR as well.

Reply:
Moved to V(2)